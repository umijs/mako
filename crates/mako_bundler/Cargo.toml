[package]
name = "mako_bundler"
version = "0.0.0"
edition = "2021"
repository = "https://github.com/umijs/mako"
license = "MIT"

[dependencies]
maplit = "1.0.2"
tokio = { workspace = true, features = ["full"] }
swc_ecma_ast = { version = "0.102.5", features = ["rkyv-impl"] }
swc_ecma_parser = { version = "0.132.6" }
swc_common = { version = "0.30.5", features = ["concurrent"] }
swc_ecma_visit = { version = "0.88.5" }
swc_ecma_codegen = { version = "0.137.6" }
swc_ecma_transforms = { version = "0.216.4", features = [
  "module",
  "typescript",
  "react",
  "concurrent",
] }
swc_ecma_utils = { version = "0.115.8" }
swc_atoms = { version = "0.4.43" }
relative-path = "1.8.0"
regex = "1"
lazy_static = "1.4.0"
base64 = "0.21.0"
petgraph = "0.6.3"
md5 = "0.7.0"
serde = { version = "1.0.160", features = ["derive"] }
serde_json = "1.0.96"
config = "0.13.3"
nodejs-resolver = "0.0.78"
lightningcss = "1.0.0-alpha.41"
swc_css_ast = "0.136.6"
swc_css_codegen = "0.146.6"
swc_css_parser = "0.145.6"
swc_css_visit = "0.135.6"
rayon = "1.7.0"

[dev-dependencies]
insta = { version = "1.29.0", features = ["yaml"] }
