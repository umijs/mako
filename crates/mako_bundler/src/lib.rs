#![feature(box_patterns)]

use compiler::Compiler;

use crate::plugins::node_polyfill::get_node_builtins;

pub mod build;
pub mod chunk;
pub mod chunk_graph;
pub mod compiler;
pub mod config;
pub mod context;
pub mod generate;
pub mod module;
pub mod module_graph;
pub mod plugin;
pub mod plugins;
pub mod utils;

pub fn run() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() <= 1 {
        panic!("Please specify the root directory of the project");
    }

    // config
    let root = std::env::current_dir()
        .unwrap()
        .join(&args[1])
        .to_string_lossy()
        .to_string();
    let mut config = config::Config::from_literal_str(
        format!(
            r#"
{{
    "externals": {{}},
    "root": "{}",
    "entry": {{ "index": "index.tsx" }}
}}
    "#,
            root
        )
        .as_str(),
    )
    .unwrap();
    config.normalize();

    // TODO: move this to plugin
    // node polyfills
    let builtins = get_node_builtins();
    for name in builtins.iter() {
        config.externals.insert(name.to_string(), name.to_string());
    }

    // compiler_origin::run_compiler(config);
    let mut compiler = Compiler::new(config);
    compiler.run();

    println!("âœ… DONE");
}
