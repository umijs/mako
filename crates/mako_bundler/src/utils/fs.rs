use rsfs::*;
use std::io;
use std::io::prelude::*;
use std::path::Path;

#[cfg(not(feature = "memory-fs"))]
pub use rsfs::disk::FS;
#[cfg(feature = "memory-fs")]
pub use rsfs::mem::FS;

#[cfg(not(feature = "memory-fs"))]
pub use rsfs::disk::File;
#[cfg(feature = "memory-fs")]
pub use rsfs::mem::File;

pub fn read_to_string<P: AsRef<Path>>(path: P) -> io::Result<String> {
    let mut file = FS.open_file(path)?;
    let mut contents = String::new();
    file.read_to_string(&mut contents)?;
    Ok(contents)
}

pub fn write<P: AsRef<Path>, C: AsRef<[u8]>>(path: P, contents: C) -> io::Result<()> {
    fn inner(path: &Path, contents: &[u8]) -> io::Result<()> {
        let f = FS {};
        FS::create_file(&f, path)?.write_all(contents)
    }
    inner(path.as_ref(), contents.as_ref())
}

pub fn read<P: AsRef<Path>>(path: P) -> io::Result<Vec<u8>> {
    fn inner(path: &Path) -> io::Result<Vec<u8>> {
        let f = FS {};
        let mut file = FS::open_file(&f, path)?;
        let mut bytes = Vec::new();
        file.read_to_end(&mut bytes)?;
        Ok(bytes)
    }
    inner(path.as_ref())
}
