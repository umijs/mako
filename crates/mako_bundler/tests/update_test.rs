use std::fs;

use mako_bundler::{
    assert_debug_snapshot, assert_display_snapshot,
    build::{
        build::BuildParam,
        update::{UpdateResult, UpdateType},
    },
    compiler::Compiler,
    config::Config,
    plugin::plugin_driver::PluginDriver,
};
use tracing::Level;
use tracing_subscriber::{EnvFilter, FmtSubscriber};

#[tokio::test(flavor = "multi_thread")]
async fn update() {
    let mut compiler = setup_compiler("update".into());
    setup_files(
        &compiler,
        vec![
            (
                "entry.js".into(),
                r#"
(async () => {
    await import('./chunk-1.js');
})();
"#
                .into(),
            ),
            (
                "chunk-1.js".into(),
                r#"
export default async function () {
    console.log(123);
}
"#
                .into(),
            ),
        ],
    );
    test_build(&mut compiler);
    {
        let module_graph = compiler.context.module_graph.read().unwrap();
        assert_display_snapshot!(&module_graph);
    }
    // 模拟文件更新，多一个文件少一个文件
    setup_files(
        &compiler,
        vec![
            (
                "entry.js".into(),
                r#"
(async () => {
    await import('./chunk-2.js');
})();
"#
                .into(),
            ),
            (
                "chunk-2.js".into(),
                r#"
export const foo = 1;
"#
                .into(),
            ),
        ],
    );
    let result = test_update(&compiler, vec![("entry.js".into(), UpdateType::Modify)]);

    {
        let module_graph = compiler.context.module_graph.read().unwrap();
        assert_display_snapshot!(&module_graph);
    }

    assert_debug_snapshot!(&result);
}

#[allow(clippy::useless_format)]
fn setup_compiler(name: String) -> Compiler {
    let subscriber = FmtSubscriber::builder()
        .with_max_level(Level::INFO)
        .with_env_filter(
            EnvFilter::try_from_default_env()
                .unwrap_or_else(|_| EnvFilter::new("mako_bundler=debug")),
        )
        .without_time()
        .finish();

    tracing::subscriber::set_global_default(subscriber).expect("Failed to set subscriber");

    let cwd = std::env::current_dir()
        .unwrap()
        .join("tests/fixtures")
        .join(&name)
        .to_string_lossy()
        .to_string();
    let mut config = Config::from_literal_str(
        format!(
            r#"
{{
    "entry": {{
        "entry": "entry.js"
    }},
    "root": "{}",
    "externals":  {{ "test": "test" }}
}}
            "#,
            cwd,
        )
        .as_str(),
    )
    .unwrap();
    config.normalize();

    Compiler::new(config, PluginDriver::new())
}

fn test_build(compiler: &mut Compiler) {
    compiler.build(&BuildParam { files: None }).unwrap();
}

fn test_update(compiler: &Compiler, paths: Vec<(String, UpdateType)>) -> UpdateResult {
    compiler.update(paths).unwrap()
}

fn setup_files(compiler: &Compiler, extra_files: Vec<(String, String)>) {
    let cwd_path = &compiler.context.config.root;
    extra_files.into_iter().for_each(|(path, content)| {
        fs::write(cwd_path.join(path), content).unwrap();
    });
}
