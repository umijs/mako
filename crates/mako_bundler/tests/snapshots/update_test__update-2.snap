---
source: crates/mako_bundler/tests/update_test.rs
expression: "&module_graph"
---
graph
 nodes:["<CWD>/tests/fixtures/update/chunk-2.js", "<CWD>/tests/fixtures/update/entry.js"] 
 references:["<CWD>/tests/fixtures/update/entry.js -> <CWD>/tests/fixtures/update/chunk-2.js"]
